File: sdk/python/foodblock/explain.py
"""
Generate human-readable narratives from FoodBlock graphs.

Usage:
    story = await explain(bread_hash, resolve)
    # "Sourdough ($4.50). By Green Acres Bakery. Made from Flour, Water."
"""

import asyncio


async def explain(hash_val, resolve, max_depth=10):
    """
    Generate a narrative for a block and its provenance.

    Args:
        hash_val: Hash of the block to explain
        resolve: async (hash) -> block or None
        max_depth: Maximum traversal depth

    Returns:
        Human-readable string
    """
    block = await resolve(hash_val)
    if not block:
        return f'Block not found: {hash_val}'

    parts = []
    visited = set()
    await _build_narrative(block, resolve, parts, visited, 0, max_depth)
    return ' '.join(parts)


async def _build_narrative(block, resolve, parts, visited, depth, max_depth):
    if not block or depth > max_depth:
        return
    block_hash = block.get('hash', '')
    if block_hash in visited:
        return
    visited.add(block_hash)

    state = block.get('state', {})
    refs = block.get('refs', {})
    name = state.get('name') or state.get('title') or block.get('type', 'Unknown')

    # Describe the block
    if depth == 0:
        desc = name
        if 'price' in state:
            desc += f' (${state["price"]})'
        if 'rating' in state:
            desc += f' ({state["rating"]}/5)'
        parts.append(desc + '.')

    # Actor refs
    for role in ['seller', 'buyer', 'author', 'operator', 'producer']:
        ref_hash = refs.get(role)
        if ref_hash and ref_hash not in visited:
            actor = await resolve(ref_hash)
            if actor and actor.get('state', {}).get('name'):
                visited.add(ref_hash)
                if depth == 0:
                    parts.append(f'By {actor["state"]["name"]}.')

    # Input/source refs
    for role in ['inputs', 'source', 'origin', 'input']:
        if role not in refs:
            continue
        ref_val = refs[role]
        ref_hashes = ref_val if isinstance(ref_val, list) else [ref_val]
        names = []
        for h in ref_hashes:
            dep = await resolve(h)
            if dep and dep.get('state', {}).get('name'):
                dep_desc = dep['state']['name']
                dep_source = (dep.get('refs') or {}).get('seller') or (dep.get('refs') or {}).get('source')
                if dep_source:
                    source_actor = await resolve(dep_source)
                    if source_actor and source_actor.get('state', {}).get('name'):
                        dep_desc += f' ({source_actor["state"]["name"]})'
                names.append(dep_desc)
        if names:
            parts.append(f'Made from {", ".join(names)}.')

    # Certifications
    cert_refs = refs.get('certifications')
    if cert_refs:
        cert_hashes = cert_refs if isinstance(cert_refs, list) else [cert_refs]
        for h in cert_hashes:
            cert = await resolve(h)
            if cert and cert.get('state', {}).get('name'):
                cert_desc = f'Certified: {cert["state"]["name"]}'
                if cert.get('state', {}).get('valid_until'):
                    cert_desc += f' (expires {cert["state"]["valid_until"]})'
                parts.append(cert_desc + '.')

    # Tombstone
    if state.get('tombstoned'):
        parts.append('This block has been erased.')

File: sdk/swift/Sources/Explain.swift
import Foundation

// MARK: - FoodBlockExplain

/// Generate human-readable narratives from FoodBlock graphs.
/// Walks the provenance tree and renders a plain-English explanation.
///
/// Usage:
///   let story = await FoodBlockExplain.explain(hash: breadHash, resolve: myResolver)
///   // "Sourdough ($4.50). By Green Acres Bakery. Made from Organic Flour (Stone Mill)."
public enum FoodBlockExplain {

    /// Async resolver: given a hash, returns the FoodBlock or nil.
    public typealias Resolver = (String) async -> FoodBlock?

    // MARK: - Type Labels

    private static let observeLabels: [String: String] = [
        "observe.review": "review",
        "observe.certification": "certified by",
        "observe.inspection": "inspected by",
        "observe.reading": "reading from",
        "observe.scan": "scanned by",
    ]

    // MARK: - Explain

    /// Generate a narrative for a block and its provenance.
    ///
    /// Walks the block's refs (actors, inputs, certifications, update chain)
    /// and produces a plain-English summary of the block's story.
    ///
    /// - Parameters:
    ///   - hash: Hash of the block to explain.
    ///   - resolve: Async function that retrieves a FoodBlock by hash.
    ///   - maxDepth: Maximum depth to traverse (default 10).
    /// - Returns: Human-readable narrative string.
    public static func explain(
        hash: String,
        resolve: @escaping Resolver,
        maxDepth: Int = 10
    ) async -> String {
        guard let block = await resolve(hash) else {
            return "Block not found: \(hash)"
        }

        var parts: [String] = []
        var visited = Set<String>()

        await buildNarrative(
            block: block,
            resolve: resolve,
            parts: &parts,
            visited: &visited,
            depth: 0,
            maxDepth: maxDepth
        )

        return parts.joined(separator: " ")
    }

    // MARK: - Private

    private static func buildNarrative(
        block: FoodBlock,
        resolve: @escaping Resolver,
        parts: inout [String],
        visited: inout Set<String>,
        depth: Int,
        maxDepth: Int
    ) async {
        if visited.contains(block.hash) || depth > maxDepth { return }
        visited.insert(block.hash)

        let name = stateString(block, key: "name")
            ?? stateString(block, key: "title")
            ?? block.type

        // Describe the block itself (only at root depth)
        if depth == 0 {
            var desc = name
            if let price = stateNumber(block, key: "price") {
                desc += " ($\(formatNumber(price)))"
            }
            if let rating = stateNumber(block, key: "rating") {
                desc += " (\(formatNumber(rating))/5)"
            }
            parts.append(desc + ".")
        }

        // Follow key refs to build the story
        let refs = block.refs

        // Actor refs (seller, buyer, author, operator, producer)
        let actorRoles = ["seller", "buyer", "author", "operator", "producer"]
        for role in actorRoles {
            if let refValue = refs[role]?.value as? String {
                if let actor = await resolve(refValue),
                   let actorName = stateString(actor, key: "name"),
                   !visited.contains(actor.hash) {
                    visited.insert(actor.hash)
                    if depth == 0 {
                        parts.append("By \(actorName).")
                    }
                }
            }
        }

        // Input/source refs (provenance)
        let inputRoles = ["inputs", "source", "origin", "input"]
        for role in inputRoles {
            guard let refVal = refs[role]?.value else { continue }
            let refHashes: [String]
            if let s = refVal as? String {
                refHashes = [s]
            } else if let arr = refVal as? [Any] {
                refHashes = arr.compactMap { $0 as? String }
            } else {
                continue
            }

            var names: [String] = []
            for h in refHashes {
                guard let dep = await resolve(h),
                      let depName = stateString(dep, key: "name") else { continue }

                var depDesc = depName

                // Check for source actor on the dependency
                let sourceRoles = ["seller", "source", "producer"]
                for sr in sourceRoles {
                    if let sourceHash = dep.refs[sr]?.value as? String {
                        if let sourceActor = await resolve(sourceHash),
                           let sourceName = stateString(sourceActor, key: "name") {
                            depDesc += " (\(sourceName))"
                            break
                        }
                    }
                }

                names.append(depDesc)
            }

            if !names.isEmpty {
                parts.append("Made from \(names.joined(separator: ", ")).")
            }
        }

        // Certifications
        if let certVal = refs["certifications"]?.value {
            let certHashes: [String]
            if let s = certVal as? String {
                certHashes = [s]
            } else if let arr = certVal as? [Any] {
                certHashes = arr.compactMap { $0 as? String }
            } else {
                certHashes = []
            }

            for h in certHashes {
                guard let cert = await resolve(h),
                      let certName = stateString(cert, key: "name") else { continue }
                var certDesc = "Certified: \(certName)"
                if let validUntil = stateString(cert, key: "valid_until") {
                    certDesc += " (expires \(validUntil))"
                }
                parts.append(certDesc + ".")
            }
        }

        // Update chain — detect price changes
        if let updatesHash = refs["updates"]?.value as? String, !visited.contains(updatesHash) {
            if let prev = await resolve(updatesHash) {
                let prevPrice = stateNumber(prev, key: "price")
                let currPrice = stateNumber(block, key: "price")
                if let pp = prevPrice, let cp = currPrice, pp != cp {
                    parts.append("Updated from $\(formatNumber(pp)).")
                }
            }
        }

        // Tombstone
        if let tombstoned = block.state["tombstoned"]?.value as? Bool, tombstoned {
            parts.append("This block has been erased.")
        }
    }

    // MARK: - Helpers

    /// Extract a string value from block state.
    private static func stateString(_ block: FoodBlock, key: String) -> String? {
        return block.state[key]?.value as? String
    }

    /// Extract a numeric value from block state (handles Int and Double).
    private static func stateNumber(_ block: FoodBlock, key: String) -> Double? {
        let val = block.state[key]?.value
        if let d = val as? Double { return d }
        if let i = val as? Int { return Double(i) }
        return nil
    }

    /// Format a number for display: omit ".0" for whole numbers.
    private static func formatNumber(_ n: Double) -> String {
        if n == n.rounded() && !n.isInfinite && !n.isNaN {
            return String(format: "%.0f", n)
        }
        return String(n)
    }
}
